core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -432,7 +432,7 @@ private[spark] class Executor(
           setTaskFinishedAndClearInterruptStatus()
           execBackend.statusUpdate(taskId, TaskState.KILLED, ser.serialize(TaskKilled(t.reason)))
 
-        case _: InterruptedException if task.reasonIfKilled.isDefined =>
+        case _: Throwable if task.reasonIfKilled.isDefined =>
```

Comment:
That said, the existing `Throwable` case has logic to exit the JVM in case the throwable was fatal (see `Utils.isFatalError(t))` used further down).

We probably want to preserve the behavior of continuing to exit on fatal errors while still trying to issue a final "dying breath" task status. Thus maybe we can catch `NonFatal` up here, as you've proposed, and add another piece of conditional logic to handle fatal errors in the existing `Throwable` case. This adds a bit of complexity and we'd need another test case to fully cover it, but that could be done by having a task which throws `OutOfMemoryError` once interrupted.